### PR TITLE
 * NEW [supplemental/mqtt] add nni_mqtt_msg_get/set_aio API

### PR DESCRIPTION
**Pull request by JaylinYu**

### CODE REVIEW

**Comment by wanghaEMQ** [comment]
* Replace the state machine with hashtable.
* Simplify the process of sending and receiving.
* Move ping handler to transport layer from protocol layer.

**Review by wanghaEMQ — COMMENTED** [review]
Are there any more changes besides I list above?

File: CMakeLists.txt [hunk 1]
```diff
@@ -258,6 +258,8 @@ endif ()
 #  Build the tools
 
 add_subdirectory(docs/man)
+add_subdirectory(mqtt)
+add_subdirectory(mqtt_async)
```
Comment (wanghaEMQ) [inline CMakeLists.txt:262]: These two directories do not exist...

Comment (alvin1221) [reply CMakeLists.txt:262]: These two subdirectories are demos, should not be added to CMakeLists.txt.

**Review by alvin1221 — COMMENTED** [review]
reviewed

File: src/mqtt/protocol/mqtt/mqtt_client.c [hunk 2]
```diff
@@ -416,87 +304,36 @@ mqtt_pipe_recv_msgq_putq(mqtt_pipe_t *p, nni_msg *msg)
...
 		// resize to ensure we do not lost messages
 		// add option to drop messages
-		if (0 !=
-		    nni_lmq_resize(&p->recv_messages,
-		        nni_lmq_len(&p->recv_messages) * 2)) {
-			// drop the message when no memory available
-			nni_msg_free(msg);
-			return;
-		}
-		nni_lmq_put(&p->recv_messages, msg);
```
Comment (wanghaEMQ) [inline src/mqtt/protocol/mqtt/mqtt_client.c:426]: Why not put msg into lmq?

**Review by wanghaEMQ — COMMENTED** [review]
Keepalive not work ...

**Comment by JaylinYu** [comment]
just a experimental PR

**Pull request closed without merging**